exporters/otlp/lib/opentelemetry/exporters/otlp/exporter.rb

```diff
@@ -0,0 +1,252 @@
+
+          @path = path
+          @max_retry_count = retry_count
+          @tracer = OpenTelemetry.tracer_provider.tracer
+
+          @shutdown = false
+        end
+
+        # Called to export sampled {OpenTelemetry::SDK::Trace::SpanData} structs.
+        #
+        # @param [Enumerable<OpenTelemetry::SDK::Trace::SpanData>] span_data the
+        #   list of recorded {OpenTelemetry::SDK::Trace::SpanData} structs to be
+        #   exported.
+        # @return [Integer] the result of the export.
+        def export(span_data)
+          return FAILURE if @shutdown
+
+          send_bytes(encode(span_data))
+        end
+
+        # Called when {OpenTelemetry::SDK::Trace::Tracer#shutdown} is called, if
+        # this exporter is registered to a {OpenTelemetry::SDK::Trace::Tracer}
+        # object.
+        def shutdown
+          @shutdown = true
+          @http.finish if @http.started?
+        end
+
+        private
+
+        def send_bytes(bytes)
+          retry_count = 0
+          untraced do
+            request = Net::HTTP::Post.new(@path)
+            request.body = bytes
+            request.add_field('Content-Type', 'application/x-protobuf')
+            # TODO: enable gzip when https://github.com/open-telemetry/opentelemetry-collector/issues/1344 is fixed.
+            # request.add_field('Content-Encoding', 'gzip')
+  
+            @http.start unless @http.started?
+            response = @http.request(request)
+
+            case response
+            when Net::HTTPOK
+              response.body # Read and discard body
+              SUCCESS
+            when Net::HTTPServiceUnavailable, Net::HTTPTooManyRequests
+              response.body # Read and discard body
+              redo if backoff?(retry_after: response['Retry-After'], retry_count: retry_count += 1)
+              FAILURE
+            when Net::HTTPRequestTimeOut, Net::HTTPGatewayTimeOut, Net::HTTPBadGateway
+              response.body # Read and discard body
+              redo if backoff?(retry_count: retry_count += 1)
+              FAILURE
+            when Net::HTTPBadRequest, Net::HTTPClientError, Net::HTTPServerError
+              # TODO: decode the body as a google.rpc.Status Protobuf-encoded message when https://github.com/open-telemetry/opentelemetry-collector/issues/1357 is fixed.
+              response.body # Read and discard body
+              FAILURE
+            when Net::HTTPRedirection
+              @http.finish
+              handle_redirect(response['location'])
+              redo if backoff?(retry_after: 0, retry_count: retry_count += 1)
+            else
+              @http.finish
+              FAILURE
+            end
+          rescue Net::OpenTimeout, Net::ReadTimeout
+            retry if backoff?(retry_count: retry_count += 1)
+            return FAILURE
+          end
+        end
+
+        def handle_redirect(location)
+          # TODO: figure out destination and reinitialize @http and @path
```

Comment:
Need to handle redirects -- this code is c&p from Shopify's implementation, and needs to be adjusted to work in this implementation.

Reply:
you should put a limit on the redirects is about all I can recommend.

Reply:
I’m going to remove this from the current PR and figure it out in a subsequent PR. AFAIK the collector won’t redirect, though IDK whether Envoy in front of a collector pool might. In any case, this PR is large enough and late enough already.